# Pandas for Exploratory Data Analysis II 

Pandas a very useful Python library for data manipulation and exploration. We have so much more to see!

In this lesson, we'll continue exploring Pandas for EDA. Specifically: 

- Identify and handle missing values with Pandas.
- Implement groupby statements for specific segmented analysis.
- Use apply functions to clean data with Pandas.

We'll implicitly review many functions from our first Pandas lesson along the way!

## Remember the AdventureWorks Cycles Dataset?
<img align="right" src="http://lh6.ggpht.com/_XjcDyZkJqHg/TPaaRcaysbI/AAAAAAAAAFo/b1U3q-qbTjY/AdventureWorks%20Logo%5B5%5D.png?imgmax=800">

Here's the Production.Product table [data dictionary](https://www.sqldatadictionary.com/AdventureWorks2014/Production.Product.html), which is a description of the fields (columns) in the table (the .csv file we will import below):<br>
- **ProductID** - Primary key for Product records.
- **Name** - Name of the product.
- **ProductNumber** - Unique product identification number.
- **MakeFlag** - 0 = Product is purchased, 1 = Product is manufactured in-house.
- **FinishedGoodsFlag** - 0 = Product is not a salable item. 1 = Product is salable.
- **Color** - Product color.
- **SafetyStockLevel** - Minimum inventory quantity.
- **ReorderPoint** - Inventory level that triggers a purchase order or work order.
- **StandardCost** - Standard cost of the product.
- **ListPrice** - Selling price.
- **Size** - Product size.
- **SizeUnitMeasureCode** - Unit of measure for the Size column.
- **WeightUnitMeasureCode** - Unit of measure for the Weight column.
- **DaysToManufacture** - Number of days required to manufacture the product.
- **ProductLine** - R = Road, M = Mountain, T = Touring, S = Standard
- **Class** - H = High, M = Medium, L = Low
- **Style** - W = Womens, M = Mens, U = Universal
- **ProductSubcategoryID** - Product is a member of this product subcategory. Foreign key to ProductSubCategory.ProductSubCategoryID.
- **ProductModelID** - Product is a member of this product model. Foreign key to ProductModel.ProductModelID.
- **SellStartDate** - Date the product was available for sale.
- **SellEndDate** - Date the product was no longer available for sale.
- **DiscontinuedDate** - Date the product was discontinued.
- **rowguid** - ROWGUIDCOL number uniquely identifying the record. Used to support a merge replication sample.
- **ModifiedDate** - Date and time the record was last updated.


### Import Pandas

In [1]:
import pandas as pd
import numpy as np # used for linear algebra and random sampling
# used for plotting charts within the notebook (instead of a separate window)
%matplotlib inline

### Read in the dataset

We are using the `read_csv()` method (and the `\t` separator to specify tab-delimited columns).

In [2]:
# read in the dataset
prod = pd.read_csv('../data/Production.Product.csv', sep='\t')

In [3]:
# let's check out the first 3 rows again, for old time's sake
prod.head(3)

ProductID             Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
0          1  Adjustable Race       AR-5381         0                  0   
1          2     Bearing Ball       BA-8327         0                  0   
2          3  BB Ball Bearing       BE-2349         1                  0   

  Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
0   NaN              1000           750           0.0        0.0  ...   
1   NaN              1000           750           0.0        0.0  ...   
2   NaN               800           600           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
0         NaN   NaN   NaN                   NaN             NaN   
1         NaN   NaN   NaN                   NaN             NaN   
2         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
0  2008-04-30 00:00:00         NaN              NaN   
1  2008-04-30 00:00:00         NaN              NaN   
2  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

# head
# shape
# describe
# info
# columns

In [4]:
prod.head()

ProductID                   Name ProductNumber  MakeFlag  \
0          1        Adjustable Race       AR-5381         0   
1          2           Bearing Ball       BA-8327         0   
2          3        BB Ball Bearing       BE-2349         1   
3          4  Headset Ball Bearings       BE-2908         0   
4        316                  Blade       BL-2036         1   

   FinishedGoodsFlag Color  SafetyStockLevel  ReorderPoint  StandardCost  \
0                  0   NaN              1000           750           0.0   
1                  0   NaN              1000           750           0.0   
2                  0   NaN               800           600           0.0   
3                  0   NaN               800           600           0.0   
4                  0   NaN               800           600           0.0   

   ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
0        0.0  ...         NaN   NaN   NaN                   NaN   
1        0.0  ...         NaN   NaN   NaN                   NaN   
2        0.0  ...         NaN   NaN   NaN                   NaN   
3        0.0  ...         NaN   NaN   NaN                   NaN   
4        0.0  ...         NaN   NaN   NaN                   NaN   

   ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
0             NaN  2008-04-30 00:00:00         NaN              NaN   
1             NaN  2008-04-30 00:00:00         NaN              NaN   
2             NaN  2008-04-30 00:00:00         NaN              NaN   
3             NaN  2008-04-30 00:00:00         NaN              NaN   
4             NaN  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  
3  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}  2014-02-08 10:01:36.827000000  
4  {E73E9750-603B-4131-89F5-3DD15ED5FF80}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

In [5]:
prod.describe()

ProductID    MakeFlag  FinishedGoodsFlag  SafetyStockLevel  \
count  504.000000  504.000000         504.000000        504.000000   
mean   673.039683    0.474206           0.585317        535.150794   
std    229.373142    0.499830           0.493157        374.112954   
min      1.000000    0.000000           0.000000          4.000000   
25%    447.750000    0.000000           0.000000        100.000000   
50%    747.500000    0.000000           1.000000        500.000000   
75%    873.250000    1.000000           1.000000       1000.000000   
max    999.000000    1.000000           1.000000       1000.000000   

       ReorderPoint  StandardCost    ListPrice       Weight  \
count    504.000000    504.000000   504.000000   205.000000   
mean     401.363095    258.602961   438.666250    74.069220   
std      280.584715    461.632808   773.602843   182.166588   
min        3.000000      0.000000     0.000000     2.120000   
25%       75.000000      0.000000     0.000000     2.880000   
50%      375.000000     23.372200    49.990000    17.900000   
75%      750.000000    317.075825   564.990000    27.350000   
max      750.000000   2171.294200  3578.270000  1050.000000   

       DaysToManufacture  ProductSubcategoryID  ProductModelID  \
count         504.000000            295.000000      295.000000   
mean            1.103175             12.294915       37.444068   
std             1.492616              9.860135       34.025442   
min             0.000000              1.000000        1.000000   
25%             0.000000              2.000000       11.000000   
50%             1.000000             12.000000       26.000000   
75%             1.000000             17.000000       48.500000   
max             4.000000             37.000000      128.000000   

       DiscontinuedDate  
count               0.0  
mean                NaN  
std                 NaN  
min                 NaN  
25%                 NaN  
50%                 NaN  
75%                 NaN  
max                 NaN

In [6]:
prod.shape

(504, 25)

In [7]:
prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 504 entries, 0 to 503
Data columns (total 25 columns):
ProductID                504 non-null int64
Name                     504 non-null object
ProductNumber            504 non-null object
MakeFlag                 504 non-null int64
FinishedGoodsFlag        504 non-null int64
Color                    256 non-null object
SafetyStockLevel         504 non-null int64
ReorderPoint             504 non-null int64
StandardCost             504 non-null float64
ListPrice                504 non-null float64
Size                     211 non-null object
SizeUnitMeasureCode      176 non-null object
WeightUnitMeasureCode    205 non-null object
Weight                   205 non-null float64
DaysToManufacture        504 non-null int64
ProductLine              278 non-null object
Class                    247 non-null object
Style                    211 non-null object
ProductSubcategoryID     295 non-null float64
ProductModelID           295 non-null floa

In [8]:
prod.columns

Index(['ProductID', 'Name', 'ProductNumber', 'MakeFlag', 'FinishedGoodsFlag',
       'Color', 'SafetyStockLevel', 'ReorderPoint', 'StandardCost',
       'ListPrice', 'Size', 'SizeUnitMeasureCode', 'WeightUnitMeasureCode',
       'Weight', 'DaysToManufacture', 'ProductLine', 'Class', 'Style',
       'ProductSubcategoryID', 'ProductModelID', 'SellStartDate',
       'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate'],
      dtype='object')

### Reset our index (like last time)

Let's bring our `ProductID` column into the index since it's the PK (primary key) of our table and that's where PKs belong as a best practice.

In [9]:
prod.head()

ProductID                   Name ProductNumber  MakeFlag  \
0          1        Adjustable Race       AR-5381         0   
1          2           Bearing Ball       BA-8327         0   
2          3        BB Ball Bearing       BE-2349         1   
3          4  Headset Ball Bearings       BE-2908         0   
4        316                  Blade       BL-2036         1   

   FinishedGoodsFlag Color  SafetyStockLevel  ReorderPoint  StandardCost  \
0                  0   NaN              1000           750           0.0   
1                  0   NaN              1000           750           0.0   
2                  0   NaN               800           600           0.0   
3                  0   NaN               800           600           0.0   
4                  0   NaN               800           600           0.0   

   ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
0        0.0  ...         NaN   NaN   NaN                   NaN   
1        0.0  ...         NaN   NaN   NaN                   NaN   
2        0.0  ...         NaN   NaN   NaN                   NaN   
3        0.0  ...         NaN   NaN   NaN                   NaN   
4        0.0  ...         NaN   NaN   NaN                   NaN   

   ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
0             NaN  2008-04-30 00:00:00         NaN              NaN   
1             NaN  2008-04-30 00:00:00         NaN              NaN   
2             NaN  2008-04-30 00:00:00         NaN              NaN   
3             NaN  2008-04-30 00:00:00         NaN              NaN   
4             NaN  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  
3  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}  2014-02-08 10:01:36.827000000  
4  {E73E9750-603B-4131-89F5-3DD15ED5FF80}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

In [10]:
# repalace auto-generated index with the ProductID column
prod.set_index('ProductID', inplace=True)

## Handling missing data

Recall missing data is a systemic, challenging problem for data scientists. Imagine conducting a poll, but some of the data gets lost, or you run out of budget and can't complete it! 😮<br><br>

"Handling missing data" itself is a broad topic. We'll focus on two components:

- Using Pandas to identify we have missing data
- Strategies to fill in missing data (known in the business as `imputing`)
- Filling in missing data with Pandas


### Identifying missing data

Before *handling*, we must identify we're missing data at all!

We have a few ways to explore missing data, and they are reminiscient of our Boolean filters.

In [11]:
# True when data isn't missing
prod.notnull()

Name  ProductNumber  MakeFlag  FinishedGoodsFlag  Color  \
ProductID                                                            
1          True           True      True               True  False   
2          True           True      True               True  False   
3          True           True      True               True  False   
4          True           True      True               True  False   
316        True           True      True               True  False   
...         ...            ...       ...                ...    ...   
995        True           True      True               True  False   
996        True           True      True               True  False   
997        True           True      True               True   True   
998        True           True      True               True   True   
999        True           True      True               True   True   

           SafetyStockLevel  ReorderPoint  StandardCost  ListPrice   Size  \
ProductID                                                                   
1                      True          True          True       True  False   
2                      True          True          True       True  False   
3                      True          True          True       True  False   
4                      True          True          True       True  False   
316                    True          True          True       True  False   
...                     ...           ...           ...        ...    ...   
995                    True          True          True       True  False   
996                    True          True          True       True  False   
997                    True          True          True       True   True   
998                    True          True          True       True   True   
999                    True          True          True       True   True   

           ...  ProductLine  Class  Style  ProductSubcategoryID  \
ProductID  ...                                                    
1          ...        False  False  False                 False   
2          ...        False  False  False                 False   
3          ...        False  False  False                 False   
4          ...        False  False  False                 False   
316        ...        False  False  False                 False   
...        ...          ...    ...    ...                   ...   
995        ...        False   True  False                  True   
996        ...        False   True  False                  True   
997        ...         True   True   True                  True   
998        ...         True   True   True                  True   
999        ...         True   True   True                  True   

           ProductModelID  SellStartDate  SellEndDate  DiscontinuedDate  \
ProductID                                                                 
1                   False           True        False             False   
2                   False           True        False             False   
3                   False           True        False             False   
4                   False           True        False             False   
316                 False           True        False             False   
...                   ...            ...          ...               ...   
995                  True           True        False             False   
996                  True           True        False             False   
997                  True           True        False             False   
998                  True           True        False             False   
999                  True           True        False             False   

           rowguid  ModifiedDate  
ProductID                         
1             True          True  
2             True          True  
3             True          True  
4             True          True  
316           True          True  
...  

In [12]:
# True when data is missing
prod.notnull().sum()

Name                     504
ProductNumber            504
MakeFlag                 504
FinishedGoodsFlag        504
Color                    256
SafetyStockLevel         504
ReorderPoint             504
StandardCost             504
ListPrice                504
Size                     211
SizeUnitMeasureCode      176
WeightUnitMeasureCode    205
Weight                   205
DaysToManufacture        504
ProductLine              278
Class                    247
Style                    211
ProductSubcategoryID     295
ProductModelID           295
SellStartDate            504
SellEndDate               98
DiscontinuedDate           0
rowguid                  504
ModifiedDate             504
dtype: int64

Now, we may want to see null values in aggregate. We can use `sum()` to sum down a given column

In [13]:
# here is a quick and dirty way to do it
prod.isnull().sum()

Name                       0
ProductNumber              0
MakeFlag                   0
FinishedGoodsFlag          0
Color                    248
SafetyStockLevel           0
ReorderPoint               0
StandardCost               0
ListPrice                  0
Size                     293
SizeUnitMeasureCode      328
WeightUnitMeasureCode    299
Weight                   299
DaysToManufacture          0
ProductLine              226
Class                    257
Style                    293
ProductSubcategoryID     209
ProductModelID           209
SellStartDate              0
SellEndDate              406
DiscontinuedDate         504
rowguid                    0
ModifiedDate               0
dtype: int64

Look! We've found missing values!

How could this missing data be problematic for our analysis?

### Understanding missing data

Finding missing data is the easy part! Determining way to do next is more complicated.

Typically, we are most interested in knowing **why** we are missing data. Once we know what 'type of missingness' we have (the source of missing data), we can proceed effectively.

Let's first quantify how much data we are missing. Here is another implementation of `prod.isnull().sum()`, only wrapped with a `DataFrame` and some labels to make it a little more user-friendly:

In [14]:
# or we can make things pretty as follows
null_df = pd.DataFrame(prod.isnull().sum(), columns=['Count of Nulls'])
null_df.index.name = 'Column'
null_df.sort_values(['Count of Nulls'], ascending=False).head(10)

Count of Nulls
Column                               
DiscontinuedDate                  504
SellEndDate                       406
SizeUnitMeasureCode               328
Weight                            299
WeightUnitMeasureCode             299
Size                              293
Style                             293
Class                             257
Color                             248
ProductLine                       226

### Filling in missing data

How we fill in data depends largely on why it is missing (types of missingness) and what sampling we have available to us.

We may:

- Delete missing data altogether
- Fill in missing data with:
    - The average of the column
    - The median of the column
    - A predicted amount based on other factors
- Collect more data:
    - Resample the population
    - Followup with the authority providing data that is missing


In [15]:
prod["Color"].value_counts()

Black           93
Silver          43
Red             38
Yellow          36
Blue            26
Multi            8
Silver/Black     7
White            4
Grey             1
Name: Color, dtype: int64

In our case, let's focus on handling missing values in `Color`. Let's get a count of the unique values in that column. We will need to use the `dropna=False` kwarg, otherwise the `pd.Series.value_counts()` method will not count `NaN` (null) values.

In [16]:
# let's get a value count with the nulls included
prod['Color'].value_counts(dropna=False)

NaN             248
Black            93
Silver           43
Red              38
Yellow           36
Blue             26
Multi             8
Silver/Black      7
White             4
Grey              1
Name: Color, dtype: int64

Ahoy! We have 248 nulls!

Option 1: Drop the missing values.

In [17]:
# drops rows where any row has a missing value - this does not happen *in place*, so we are not actually dropping
prod['Color'].dropna(inplace=False)

ProductID
317     Black
318     Black
319     Black
320    Silver
321    Silver
        ...  
992     Black
993     Black
997     Black
998     Black
999     Black
Name: Color, Length: 256, dtype: object

In [18]:
prod.dropna(how='any', subset=['Color'], inplace=False)

Column                       Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                      
317                   LL Crankarm       CA-5965         0                  0   
318                   ML Crankarm       CA-6738         0                  0   
319                   HL Crankarm       CA-7457         0                  0   
320               Chainring Bolts       CB-2903         0                  0   
321                 Chainring Nut       CN-6137         0                  0   
...                           ...           ...       ...                ...   
992        Mountain-500 Black, 48    BK-M18B-48         1                  1   
993        Mountain-500 Black, 52    BK-M18B-52         1                  1   
997            Road-750 Black, 44    BK-R19B-44         1                  1   
998            Road-750 Black, 48    BK-R19B-48         1                  1   
999            Road-750 Black, 52    BK-R19B-52         1                  1   

Column      Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                    
317         Black               500           375        0.0000       0.00   
318         Black               500           375        0.0000       0.00   
319         Black               500           375        0.0000       0.00   
320        Silver              1000           750        0.0000       0.00   
321        Silver              1000           750        0.0000       0.00   
...           ...               ...           ...           ...        ...   
992         Black               100            75      294.5797     539.99   
993         Black               100            75      294.5797     539.99   
997         Black               100            75      343.6496     539.99   
998         Black               100            75      343.6496     539.99   
999         Black               100            75      343.6496     539.99   

Column    Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
317        NaN  ...         NaN    L     NaN                   NaN   
318        NaN  ...         NaN    M     NaN                   NaN   
319        NaN  ...         NaN   NaN    NaN                   NaN   
320        NaN  ...         NaN   NaN    NaN                   NaN   
321        NaN  ...         NaN   NaN    NaN                   NaN   
...        ...  ...         ...   ...    ...                   ...   
992         48  ...          M     L      U                    1.0   
993         52  ...          M     L      U                    1.0   
997         44  ...          R     L      U                    2.0   
998         48  ...          R     L      U                    2.0   
999         52  ...          R     L      U                    2.0   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
317                  NaN  2008-04-30 00:00:00         NaN               NaN   
318                  NaN  2008-04-30 00:00:00         NaN               NaN   
319                  NaN  2008-04-30 00:00:00         NaN               NaN   
320                  NaN  2008-04-30 00:00:00         NaN               NaN   
321                  NaN  2008-04-30 00:00:00         NaN               NaN   
...                  ...                  ...         ...               ...   
992                 23.0  2013-05-30 00:00:00         NaN               NaN   
993                 23.0  2013-05-30 00:00:00         NaN               NaN   
997                 31.0  2013-05-30 00:00:00         NaN               NaN   
998                 31.0  2013-05-30 00:00:00         NaN               NaN   
999                 31.0  2013-05-30 00:00:00         NaN               NaN   

Colum

In [19]:
prod['Color'] = prod['Color'].fillna(value=)

SyntaxError: invalid syntax (<ipython-input-19-7a561eb131ef>, line 1)

In [20]:
prod.dropna()

Empty DataFrame
Columns: [Name, ProductNumber, MakeFlag, FinishedGoodsFlag, Color, SafetyStockLevel, ReorderPoint, StandardCost, ListPrice, Size, SizeUnitMeasureCode, WeightUnitMeasureCode, Weight, DaysToManufacture, ProductLine, Class, Style, ProductSubcategoryID, ProductModelID, SellStartDate, SellEndDate, DiscontinuedDate, rowguid, ModifiedDate]
Index: []

[0 rows x 24 columns]

**Important!** `pd.DataFrame.dropna()` and `pd.Series.dropna()` are very versatile! Let's look at the docs (Series is similar):

```python
Signature: pd.DataFrame.dropna(self, axis=0, how='any', thresh=None, subset=None, inplace=False)
Docstring:
Remove missing values.

See the :ref:`User Guide <missing_data>` for more on which values are
considered missing, and how to work with missing data.

Parameters
----------
axis : {0 or 'index', 1 or 'columns'}, default 0
    Determine if rows or columns which contain missing values are
    removed.

    * 0, or 'index' : Drop rows which contain missing values.
    * 1, or 'columns' : Drop columns which contain missing value.

    .. deprecated:: 0.23.0: Pass tuple or list to drop on multiple
    axes.
how : {'any', 'all'}, default 'any'
    Determine if row or column is removed from DataFrame, when we have
    at least one NA or all NA.

    * 'any' : If any NA values are present, drop that row or column.
    * 'all' : If all values are NA, drop that row or column.
thresh : int, optional
    Require that many non-NA values.
subset : array-like, optional
    Labels along other axis to consider, e.g. if you are dropping rows
    these would be a list of columns to include.
inplace : bool, default False
    If True, do operation inplace and return None.
```

**how**: This tells us if we want to remove a row if _any_ of the columns have a null, or _all_ of the columns have a null.<br>
**subset**: We can input an array here, like `['Color', 'Size', 'Weight']`, and it will only consider nulls in those columns. This is very useful!<br>
**inplace**: This is if you want to mutate (change) the source dataframe. Default is `False`, so it will return a _copy_ of the source dataframe.

To accomplish the same thing, but implement it on our entire dataframe, we can do the following:

In [21]:
# drops all nulls from the Color column, but returns the entire dataframe instead of just the Color column
prod.dropna(how='any', subset=['Color'], inplace=False).shape

(256, 24)

Option 2: Fill in missing values

Traditionally, we fill missing data with a median, average, or mode (most frequently occurring). For `Color`, let's replace the nulls with the string value `NoColor`.

Let's first look at the way we'd do it with a single column, using the `pd.Series.fillna()` method:

In [22]:
prod['Color'].fillna(value='NoColor').head(10)

ProductID
1      NoColor
2      NoColor
3      NoColor
4      NoColor
316    NoColor
317      Black
318      Black
319      Black
320     Silver
321     Silver
Name: Color, dtype: object

Now let's see how we'd do it to the whole dataframe, using the `pd.DataFrame.fillna()` method. Notice the similar API between the methods with the `value` kwarg. Good congruent design, pandas development team! The full dataframe is returned, and not just a column.

In [23]:
prod.fillna(value={'Size': 'Unknown Size'})

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column             Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID                ...                                                  
1          Unknown Size  ...         NaN   NaN    NaN                   NaN   
2          Unknown Size  ...         NaN   NaN    NaN                   NaN   
3          Unknown Size  ...         NaN   NaN    NaN                   NaN   
4          Unknown Size  ...         NaN   NaN    NaN                   NaN   
316        Unknown Size  ...         NaN   NaN    NaN                   NaN   
...                 ...  ...         ...   ...    ...                   ...   
995        Unknown Size  ...         NaN    M     NaN                   5.0   
996        Unknown Size  ...         NaN    H     NaN                   5.0   
997                  44  ...          R     L      U                    2.0   
998                  48  ...          R     L      U                    2.0   
999                  52  ...          R     L      U                    2.0   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   
4                    NaN  2008-04-30 00:00:00         NaN               NaN   
316                  NaN  2008-04-30 00:00:00         NaN               NaN   
...                  ...                  ...         ...               ...   
995                 96.0  2013-05-30 00:00:00         NaN               NaN   
996                 97.0  2013-05-30 00:00:00         NaN               NaN   
997                 31.0  2013-05-30 00:00:00         NaN               NaN   
998                 31.0  2013-05-30 00:00:00         NaN               N

In [24]:
prod.fillna(value={'Weight': prod.Weight.mean()})

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column    Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
1          NaN  ...         NaN   NaN    NaN                   NaN   
2          NaN  ...         NaN   NaN    NaN                   NaN   
3          NaN  ...         NaN   NaN    NaN                   NaN   
4          NaN  ...         NaN   NaN    NaN                   NaN   
316        NaN  ...         NaN   NaN    NaN                   NaN   
...        ...  ...         ...   ...    ...                   ...   
995        NaN  ...         NaN    M     NaN                   5.0   
996        NaN  ...         NaN    H     NaN                   5.0   
997         44  ...          R     L      U                    2.0   
998         48  ...          R     L      U                    2.0   
999         52  ...          R     L      U                    2.0   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   
4                    NaN  2008-04-30 00:00:00         NaN               NaN   
316                  NaN  2008-04-30 00:00:00         NaN               NaN   
...                  ...                  ...         ...               ...   
995                 96.0  2013-05-30 00:00:00         NaN               NaN   
996                 97.0  2013-05-30 00:00:00         NaN               NaN   
997                 31.0  2013-05-30 00:00:00         NaN               NaN   
998                 31.0  2013-05-30 00:00:00         NaN               NaN   
999                 31.0  2013-05-30 00:00:00         NaN               NaN   

Column                         

But wait! There's more! We can reference any other data or formulas we want with the imputation (the value we fill the nulls with). This is very handy if you want to impute with the average or median of that column... or even another column altogether! Here is an example where we will the nulls of `Color` with the average value from the `ListPrice` column. This has no practical value in this application, but immense value in other applications.

In [25]:
prod.fillna(value={'Color': prod['ListPrice'].mean()}).head(3)

Column                Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                               
1          Adjustable Race       AR-5381         0                  0   
2             Bearing Ball       BA-8327         0                  0   
3          BB Ball Bearing       BE-2349         1                  0   

Column       Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                     
1          438.666              1000           750           0.0        0.0   
2          438.666              1000           750           0.0        0.0   
3          438.666               800           600           0.0        0.0   

Column    Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
1          NaN  ...         NaN   NaN    NaN                   NaN   
2          NaN  ...         NaN   NaN    NaN                   NaN   
3          NaN  ...         NaN   NaN    NaN                   NaN   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   

Column                                    rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   

Column                      ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  

[3 rows x 24 columns]

In [38]:
prod.notnull().sum()

Column
Name                     504
ProductNumber            504
MakeFlag                 504
FinishedGoodsFlag        504
Color                    256
SafetyStockLevel         504
ReorderPoint             504
StandardCost             504
ListPrice                504
Size                     211
SizeUnitMeasureCode      176
WeightUnitMeasureCode    205
Weight                   205
DaysToManufacture        504
ProductLine              278
Class                    247
Style                    211
ProductSubcategoryID     295
ProductModelID           295
SellStartDate            504
SellEndDate               98
DiscontinuedDate           0
rowguid                  504
ModifiedDate             504
dtype: int64

In [56]:
from datetime import datetime, timezone

prod.fillna(value={'Color': 'Unknown Color'})
prod.drop('DiscontinuedDate', axis=1)
prod.fillna(value={'SellEndDate': datetime.now().strftime('%Y-%m-%d %H:%M:%S')})
prod.fillna(value={'SizeUnitMeasureCode': 'CM'})

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column    Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
1          NaN  ...         NaN   NaN    NaN                   NaN   
2          NaN  ...         NaN   NaN    NaN                   NaN   
3          NaN  ...         NaN   NaN    NaN                   NaN   
4          NaN  ...         NaN   NaN    NaN                   NaN   
316        NaN  ...         NaN   NaN    NaN                   NaN   
...        ...  ...         ...   ...    ...                   ...   
995        NaN  ...         NaN    M     NaN                   5.0   
996        NaN  ...         NaN    H     NaN                   5.0   
997         44  ...          R     L      U                    2.0   
998         48  ...          R     L      U                    2.0   
999         52  ...          R     L      U                    2.0   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   
4                    NaN  2008-04-30 00:00:00         NaN               NaN   
316                  NaN  2008-04-30 00:00:00         NaN               NaN   
...                  ...                  ...         ...               ...   
995                 96.0  2013-05-30 00:00:00         NaN               NaN   
996                 97.0  2013-05-30 00:00:00         NaN               NaN   
997                 31.0  2013-05-30 00:00:00         NaN               NaN   
998                 31.0  2013-05-30 00:00:00         NaN               NaN   
999                 31.0  2013-05-30 00:00:00         NaN               NaN   

Column                         

In [57]:
prod.WeightUnitMeasureCode.unique()

array(['LB '], dtype=object)

In [58]:
prod.replace(to_replace='LB', value='Pounds')

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column    Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
1          NaN  ...         NaN   NaN    NaN                   NaN   
2          NaN  ...         NaN   NaN    NaN                   NaN   
3          NaN  ...         NaN   NaN    NaN                   NaN   
4          NaN  ...         NaN   NaN    NaN                   NaN   
316        NaN  ...         NaN   NaN    NaN                   NaN   
...        ...  ...         ...   ...    ...                   ...   
995        NaN  ...         NaN    M     NaN                   5.0   
996        NaN  ...         NaN    H     NaN                   5.0   
997         44  ...          R     L      U                    2.0   
998         48  ...          R     L      U                    2.0   
999         52  ...          R     L      U                    2.0   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   
4                    NaN  2008-04-30 00:00:00         NaN               NaN   
316                  NaN  2008-04-30 00:00:00         NaN               NaN   
...                  ...                  ...         ...               ...   
995                 96.0  2013-05-30 00:00:00         NaN               NaN   
996                 97.0  2013-05-30 00:00:00         NaN               NaN   
997                 31.0  2013-05-30 00:00:00         NaN               NaN   
998                 31.0  2013-05-30 00:00:00         NaN               NaN   
999                 31.0  2013-05-30 00:00:00         NaN               NaN   

Column                         

In [59]:
prod.ProductSubcategoryID.unique()

array([nan, 14., 31., 23., 19., 21., 12.,  2.,  1., 10., 11.,  4., 17.,
       22., 35., 34., 36., 33., 24., 18., 20., 25., 28., 37., 26., 29.,
       30., 27., 32., 16.,  9.,  6., 15., 13.,  8.,  7.,  3.,  5.])

In [60]:
prod

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column    Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
1          NaN  ...         NaN   NaN    NaN                   NaN   
2          NaN  ...         NaN   NaN    NaN                   NaN   
3          NaN  ...         NaN   NaN    NaN                   NaN   
4          NaN  ...         NaN   NaN    NaN                   NaN   
316        NaN  ...         NaN   NaN    NaN                   NaN   
...        ...  ...         ...   ...    ...                   ...   
995        NaN  ...         NaN    M     NaN                   5.0   
996        NaN  ...         NaN    H     NaN                   5.0   
997         44  ...          R     L      U                    2.0   
998         48  ...          R     L      U                    2.0   
999         52  ...          R     L      U                    2.0   

Column    ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   
4                    NaN  2008-04-30 00:00:00         NaN               NaN   
316                  NaN  2008-04-30 00:00:00         NaN               NaN   
...                  ...                  ...         ...               ...   
995                 96.0  2013-05-30 00:00:00         NaN               NaN   
996                 97.0  2013-05-30 00:00:00         NaN               NaN   
997                 31.0  2013-05-30 00:00:00         NaN               NaN   
998                 31.0  2013-05-30 00:00:00         NaN               NaN   
999                 31.0  2013-05-30 00:00:00         NaN               NaN   

Column                         

In [61]:
prod.columns

Index(['Name', 'ProductNumber', 'MakeFlag', 'FinishedGoodsFlag', 'Color',
       'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size',
       'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight',
       'DaysToManufacture', 'ProductLine', 'Class', 'Style',
       'ProductSubcategoryID', 'ProductModelID', 'SellStartDate',
       'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate'],
      dtype='object', name='Column')

They're gone! Important points:

- Don't forget to use the `inplace=True` kwarg to mutate the source dataframe (i.e. 'save changes'). 
- It is helpful to not use `inplace=True` initially to ensure your code/logic is correct, prior to making permanent changes.

## Groupby Statements

In Pandas, groupby statements are similar to pivot tables in that they allow us to segment our population to a specific subset.

For example, if we want to know the average number of bottles sold and pack sizes per city, a groupby statement would make this task much more straightforward.


To think how a groupby statement works, think about it like this:

- **Split:** Separate our DataFrame by a specific attribute, for example, group by `Color`
- **Combine:** Put our DataFrame back together and return some _aggregated_ metric, such as the `sum`, `count`, or `max`.

![](http://i.imgur.com/yjNkiwL.png)

Let's try it out!

Let's group by `Color`, and get a count of products for each color.

In [63]:
# group by Color, giving the number of products of each %color
prod.groupby(by=['Color', 'Style']).count()

Column        Name  ProductNumber  MakeFlag  FinishedGoodsFlag  \
Color  Style                                                     
Black  M         4              4         4                  4   
       U        61             61        61                 61   
       W         6              6         6                  6   
Blue   U        25             25        25                 25   
Multi  M         3              3         3                  3   
       U         5              5         5                  5   
Red    U        37             37        37                 37   
Silver U        22             22        22                 22   
       W         8              8         8                  8   
White  U         4              4         4                  4   
Yellow U        22             22        22                 22   
       W        14             14        14                 14   

Column        SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  Size  \
Color  Style                                                                  
Black  M                     4             4             4          4     4   
       U                    61            61            61         61    61   
       W                     6             6             6          6     6   
Blue   U                    25            25            25         25    25   
Multi  M                     3             3             3          3     3   
       U                     5             5             5          5     4   
Red    U                    37            37            37         37    37   
Silver U                    22            22            22         22    22   
       W                     8             8             8          8     8   
White  U                     4             4             4          4     4   
Yellow U                    22            22            22         22    22   
       W                    14            14            14         14    14   

Column        SizeUnitMeasureCode  ...  DaysToManufacture  ProductLine  Class  \
Color  Style                       ...                                          
Black  M                        0  ...                  4            4      0   
       U                       55  ...                 61           61     55   
       W                        0  ...                  6            6      0   
Blue   U                       22  ...                 25           25     22   
Multi  M                        0  ...                  3            3      0   
       U                        0  ...                  5            5      0   
Red    U                       37  ...                 37           37     37   
Silver U                       22  ...                 22           22     22   
       W                        8  ...                  8            8      8   
White  U                        0  ...                  4            4      0   
Yellow U                       18  ...                 22           22     18   
       W                       14  ...                 14           14     14   

Column        ProductSubcategoryID  ProductModelID  SellStartDate  \
Color  Style                                                        
Black  M                         4               4              4   
       U                        61              61             61   
       W                         6               6              6   
Blue   U                        25              25             25   
Multi  M                         3               3              3   
       U                         5               5              5   
Red    U                        37              37             37   
Silver U                        22              22             22   
       W                         8               8              8   
White  U                         4               4              4   
Yello

In [65]:
prod.groupby(by=['Color']).mean()

Column        MakeFlag  FinishedGoodsFlag  SafetyStockLevel  ReorderPoint  \
Color                                                                       
Black         0.774194           0.956989        285.677419    214.258065   
Blue          0.846154           1.000000        223.692308    167.769231   
Grey          0.000000           1.000000          4.000000      3.000000   
Multi         0.000000           1.000000          4.000000      3.000000   
Red           0.973684           1.000000        276.421053    207.315789   
Silver        0.744186           0.837209        388.465116    291.348837   
Silver/Black  0.000000           1.000000        500.000000    375.000000   
White         0.000000           1.000000          4.000000      3.000000   
Yellow        0.888889           1.000000        244.888889    183.666667   

Column        StandardCost    ListPrice      Weight  DaysToManufacture  \
Color                                                                    
Black           415.446239   725.121075  123.021250           1.774194   
Blue            567.159477   923.679231   17.610909           2.346154   
Grey             51.562500   125.000000         NaN           0.000000   
Multi            34.031775    59.865000         NaN           0.000000   
Red             858.178055  1401.950000   10.173784           2.552632   
Silver          466.512751   850.305349   40.579412           2.000000   
Silver/Black     28.424286    64.018571  187.333333           1.000000   
White             3.379300     9.245000         NaN           0.000000   
Yellow          597.434781   959.091389   13.820625           2.416667   

Column        ProductSubcategoryID  ProductModelID  DiscontinuedDate  
Color                                                                 
Black                    11.292135       29.247191               NaN  
Blue                     11.115385       21.923077               NaN  
Grey                     35.000000      120.000000               NaN  
Multi                    19.625000       10.250000               NaN  
Red                       8.131579       19.789474               NaN  
Silver                    7.027778       31.361111               NaN  
Silver/Black             13.000000       64.142857               NaN  
White                    23.000000       21.000000               NaN  
Yellow                    9.527778       23.888889               NaN

What do we notice about this output? Are all columns the same? Why or why not?

We can see that the `.count()` method excludes nulls, and there is no way to change this with the current implementation:
```python
Signature: .count()
Docstring: Compute count of group, excluding missing values 
File:      ~/miniconda3/envs/ga/lib/python3.7/site-packages/pandas/core/groupby/groupby.py
Type:      method
```

As a best practice, you should either:
- fill in nulls prior to your .count(), or
- use the PK (primary key) of the table, which is guaranteed non-null

In [46]:
# here we can use 'x' as a dummy placeholder for nulls, simply to get consistent counts for all columns
prod.groupby('Color', as_index=False).count()['Color']

0           Black
1            Blue
2            Grey
3           Multi
4             Red
5          Silver
6    Silver/Black
7           White
8          Yellow
Name: Color, dtype: object

Let's find out the most expensive price for an item, by `Color`:

In [69]:
prod[['Color', 'ListPrice']].groupby(by='Color').mean().sort_values('ListPrice', ascending=False)

Column          ListPrice
Color                    
Red           1401.950000
Yellow         959.091389
Blue           923.679231
Silver         850.305349
Black          725.121075
Grey           125.000000
Silver/Black    64.018571
Multi           59.865000
White            9.245000

We can also do multi-level groupbys. This is referred to as a `Multiindex` dataframe. Here, we can see the following fields in a nested group by, with a count of Name (with nulls filled!); effectively giving us a count of the number of products for every unique Class/Style combination:

- Class - H = High, M = Medium, L = Low
- Style - W = Womens, M = Mens, U = Universal

In [70]:
prod.fillna(value={'Name': 'x'}).groupby(by=['Class', 'Style']).count()[['Name']]

Column       Name
Class Style      
H     U        64
L     U        68
M     U        22
      W        22

We can also use the `.agg()` method with multiple arguments, to simulate a `.describe()` method like we used before:

# prod.groupby.funtion

In [73]:
prod.groupby(['Color']).mean()
prod.groupby(['Color', 'DaysToManufacture']).mean()

Column                          MakeFlag  FinishedGoodsFlag  SafetyStockLevel  \
Color        DaysToManufacture                                                  
Black        0                    0.0000           0.809524        122.285714   
             1                    1.0000           1.000000        500.000000   
             2                    1.0000           1.000000        500.000000   
             4                    1.0000           1.000000        100.000000   
Blue         0                    0.0000           1.000000          4.000000   
             1                    1.0000           1.000000        500.000000   
             4                    1.0000           1.000000        100.000000   
Grey         0                    0.0000           1.000000          4.000000   
Multi        0                    0.0000           1.000000          4.000000   
Red          0                    0.0000           1.000000          4.000000   
             1                    1.0000           1.000000        500.000000   
             4                    1.0000           1.000000        100.000000   
Silver       0                    0.0000           0.125000        700.500000   
             1                    0.8125           1.000000        500.000000   
             2                    1.0000           1.000000        500.000000   
             4                    1.0000           1.000000        100.000000   
Silver/Black 1                    0.0000           1.000000        500.000000   
White        0                    0.0000           1.000000          4.000000   
Yellow       0                    0.0000           1.000000          4.000000   
             1                    1.0000           1.000000        500.000000   
             2                    1.0000           1.000000        500.000000   
             4                    1.0000           1.000000        100.000000   

Column                          ReorderPoint  StandardCost    ListPrice  \
Color        DaysToManufacture                                            
Black        0                     91.714286     17.042362    42.730000   
             1                    375.000000    294.548054   527.042051   
             2                    375.000000    353.881767   649.383333   
             4                     75.000000    857.653040  1467.871333   
Blue         0                      3.000000     21.083325    56.372500   
             1                    375.000000    378.470478   631.415556   
             4                     75.000000    865.813754  1392.879231   
Grey         0                      3.000000     51.562500   125.000000   
Multi        0                      3.000000     34.031775    59.865000   
Red          0                      3.000000     13.086300    34.990000   
             1                    375.000000    487.418859   799.204118   
             4                     75.000000   1215.577960  1982.632000   
Silver       0                    525.375000      2.570787     6.873750   
             1                    375.000000    267.664569   503.291250   
             2                    375.000000    363.723233   664.213333   
             4                     75.000000    916.604950  1653.927500   
Silver/Black 1                    375.000000     28.424286    64.018571   
White        0                      3.000000      3.379300     9.245000   
Yellow       0                      3.000000     41.572300    53.990000   
             1                    375.000000    373.077346   620.158462   
             2                    375.000000    360.942800   594.830000   
             4                     75.000000    896.134144  1425.246667   

Column                              Weight  ProductSubcategoryID  \
Color        DaysToManufacture                                     
Black        0                         NaN             22.176471   
             1                  232.484194        

In [49]:
prod.groupby(by=['Color'])['ListPrice'].agg(['count', 'mean', 'min', 'max'])

count         mean     min      max
Color                                            
Black            93   725.121075    0.00  3374.99
Blue             26   923.679231   34.99  2384.07
Grey              1   125.000000  125.00   125.00
Multi             8    59.865000    8.99    89.99
Red              38  1401.950000   34.99  3578.27
Silver           43   850.305349    0.00  3399.99
Silver/Black      7    64.018571   40.49    80.99
White             4     9.245000    8.99     9.50
Yellow           36   959.091389   53.99  2384.07

## Apply functions for column operations

Apply functions allow us to perform a complex operation across an entire columns highly efficiently.

For example, let's say we want to change our colors from a word, to just a single letter. How would we do that?

The first step is writing a function, with the argument being the value we would receive from each cell in the column. This function will mutate the input, and return the result. This result will then be _applied_ to the source dataframe (if desired).

In [77]:
prod.Weight

ProductID
1         NaN
2         NaN
3         NaN
4         NaN
316       NaN
        ...  
995    168.00
996    170.00
997     19.77
998     20.13
999     20.42
Name: Weight, Length: 504, dtype: float64

In [ ]:
# If weight>100, mark weight as High. If less than 100, mark weight as low, if Nan then marl weight as unknown

In [79]:
prod.head(1)

Column                Name ProductNumber  MakeFlag  FinishedGoodsFlag Color  \
ProductID                                                                     
1          Adjustable Race       AR-5381         0                  0   NaN   

Column     SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  ...  \
ProductID                                                                ...   
1                      1000           750           0.0        0.0  NaN  ...   

Column    ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID                                                                 
1                 NaN   NaN    NaN                   NaN            NaN   

Column           SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   

Column                                    rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   

Column                      ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  

[1 rows x 24 columns]

In [88]:
def markWeight(row):
    if row["Weight"] > 100:
        return "High"
    elif row["Weight"] <= 100:
        return "Low"
    else:
        return "UNKNOWN"
        

In [89]:
prod.apply(markWeight, axis=1)

ProductID
1      UNKNOWN
2      UNKNOWN
3      UNKNOWN
4      UNKNOWN
316    UNKNOWN
        ...   
995       High
996       High
997        Low
998        Low
999        Low
Length: 504, dtype: object

In [91]:
prod["MarkedWeight"] = prod.apply(markWeight, axis=1)

In [92]:
prod

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column    Size  ... Class Style  ProductSubcategoryID  ProductModelID  \
ProductID       ...                                                     
1          NaN  ...   NaN   NaN                   NaN             NaN   
2          NaN  ...   NaN   NaN                   NaN             NaN   
3          NaN  ...   NaN   NaN                   NaN             NaN   
4          NaN  ...   NaN   NaN                   NaN             NaN   
316        NaN  ...   NaN   NaN                   NaN             NaN   
...        ...  ...   ...   ...                   ...             ...   
995        NaN  ...    M    NaN                   5.0            96.0   
996        NaN  ...    H    NaN                   5.0            97.0   
997         44  ...    L     U                    2.0            31.0   
998         48  ...    L     U                    2.0            31.0   
999         52  ...    L     U                    2.0            31.0   

Column           SellStartDate SellEndDate DiscontinuedDate  \
ProductID                                                     
1          2008-04-30 00:00:00         NaN              NaN   
2          2008-04-30 00:00:00         NaN              NaN   
3          2008-04-30 00:00:00         NaN              NaN   
4          2008-04-30 00:00:00         NaN              NaN   
316        2008-04-30 00:00:00         NaN              NaN   
...                        ...         ...              ...   
995        2013-05-30 00:00:00         NaN              NaN   
996        2013-05-30 00:00:00         NaN              NaN   
997        2013-05-30 00:00:00         NaN              NaN   
998        2013-05-30 00:00:00         NaN              NaN   
999        2013-05-30 00:00:00         NaN              NaN   

Column                                    rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D5688

In [90]:
# get a list of all unique colors
prod['Color'].unique()

array([nan, 'Black', 'Silver', 'Red', 'White', 'Blue', 'Multi', 'Yellow',
       'Grey', 'Silver/Black'], dtype=object)

In [99]:
# create a function to convert colors to letters
def color_to_letter(color): 
    color_dict = {
        'Black': 'B',
        'Silver': 'S',
        'Red': 'R',
        'White': 'W',
        'Blue': 'L',
        'Multi': 'M',
        'Yellow': 'Y',
        'Grey': 'G',
        'Silver/Black': 'V'
    }
    
    try:
        return color_dict[color]
    except:
        return 'No Color'

In [100]:
color_to_letter('greg')

'No Color'

In [101]:
prod["ListPrice"]

ProductID
1        0.00
2        0.00
3        0.00
4        0.00
316      0.00
        ...  
995    101.24
996    121.49
997    539.99
998    539.99
999    539.99
Name: ListPrice, Length: 504, dtype: float64

In [107]:
def add_100(currentPrice):
    return currentPrice + 100

In [124]:
def safe_reorder(current_reorder):
    return int((current_reorder/3) *2)

In [125]:
prod['ReorderPoint'].apply(safe_reorder)

ProductID
1      500
2      500
3      400
4      400
316    400
      ... 
995    250
996    250
997     50
998     50
999     50
Name: ReorderPoint, Length: 504, dtype: int64

In [126]:
prod["SaferReorderAmount"] = prod['ReorderPoint'].apply(safe_reorder)

In [127]:
prod

Column                      Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

Column     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                   
1            NaN              1000           750        0.0000       0.00   
2            NaN              1000           750        0.0000       0.00   
3            NaN               800           600        0.0000       0.00   
4            NaN               800           600        0.0000       0.00   
316          NaN               800           600        0.0000       0.00   
...          ...               ...           ...           ...        ...   
995          NaN               500           375       44.9506     101.24   
996          NaN               500           375       53.9416     121.49   
997        Black               100            75      343.6496     539.99   
998        Black               100            75      343.6496     539.99   
999        Black               100            75      343.6496     539.99   

Column    Size  ... Style ProductSubcategoryID  ProductModelID  \
ProductID       ...                                              
1          NaN  ...   NaN                  NaN             NaN   
2          NaN  ...   NaN                  NaN             NaN   
3          NaN  ...   NaN                  NaN             NaN   
4          NaN  ...   NaN                  NaN             NaN   
316        NaN  ...   NaN                  NaN             NaN   
...        ...  ...   ...                  ...             ...   
995        NaN  ...   NaN                  5.0            96.0   
996        NaN  ...   NaN                  5.0            97.0   
997         44  ...    U                   2.0            31.0   
998         48  ...    U                   2.0            31.0   
999         52  ...    U                   2.0            31.0   

Column           SellStartDate SellEndDate DiscontinuedDate  \
ProductID                                                     
1          2008-04-30 00:00:00         NaN              NaN   
2          2008-04-30 00:00:00         NaN              NaN   
3          2008-04-30 00:00:00         NaN              NaN   
4          2008-04-30 00:00:00         NaN              NaN   
316        2008-04-30 00:00:00         NaN              NaN   
...                        ...         ...              ...   
995        2013-05-30 00:00:00         NaN              NaN   
996        2013-05-30 00:00:00         NaN              NaN   
997        2013-05-30 00:00:00         NaN              NaN   
998        2013-05-30 00:00:00         NaN              NaN   
999        2013-05-30 00:00:00         NaN              NaN   

Column                                    rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-

Now we can _apply_ this function to our `pd.Series` object, returning the result (which we can use to overwrite the source, if we choose).

In [98]:
prod['Color'].apply(color_to_letter).head()

ProductID
1      None
2      None
3      None
4      None
316    None
Name: Color, dtype: object

The `pd.DataFrame.apply` implementation is similar, however it effectively 'scrolls through' the columns and passes each one sequentially to your function:

```python
Objects passed to the function are Series objects whose index is
either the DataFrame's index (``axis=0``) or the DataFrame's columns
(``axis=1``).
```

It should only be used when you wish to apply the same function to all columns (or rows) of your `pd.DataFrame` object.

We can also use `pd.Series.apply()` with a **labmda expression**. This is an undeclared function and is commonly used for simple functions within the `.apply()` method. Let's use it to add $100 to our `ListPrice` column. Hey, baby needs new shoes!

In [54]:
# prior to apply method

In [55]:
# and now with 100 more dollars!

Boom! Maybe financing that new boat wasn't such a bad idea after all!

**Your turn:** Identify one other column where we may want to write a new apply function, or use the one we just created for the purposes of cleaning up our dataset.

In [56]:
0# identify a column to mutate (change)


0

In [57]:
# write a function to mutate that column (or columns) note: if using a lambda function, you can leave this blank


In [58]:
# apply that function across the whole column


## Wrap up

We've covered even more useful information! Here are the key takeaways:

- **Missing data** comes in many shapes and sizes. Before deciding how to handle it, we identify it exists. We then derive how the missingness is affecting our dataset, and make a determination about how to fill in values.

```python
# pro tip for identifying missing data
df.isnull().sum()
```

- **Grouby** statements are particularly useful for a subsection-of-interest analysis. Specifically, zooming in on one condition, and determining relevant statstics.

```python
# group by 
df.groupby('column').agg['count', 'mean', 'max', 'min']
```

- **Apply functions** help us clean values across an entire DataFrame column. They are *like* a for loop for cleaning, but many times more efficient. They follow a common pattern:
1. Write a function that works on a single value
2. Test that function on a single value
3. Apply that function to a whole column

(The most confusing part of apply functions is that we write them with *a single value* in mind, and then apply them to many single values at once.)